File: spike-1/src/spike_1/timeline.clj
(ns spike-1.timeline
  (:require [spike-1.heatmap]))

(defprotocol ITimeline
  (effects [timeline] ""))

(defrecord Timeline
  [effects]
  ITimeline
    (effects [timeline] (:effects timeline)))
  Heatmappable
    (heatmap [timeline]
             (let [effects (map heatmap (effects timeline))]
                  (reduce + effects))))
(defn timeline
  ""
  [effects]
  (->Timeline effects))

File: spike-3/src/timeline.js
import { List } from 'immutable'
import { Heatmap, heatmap } from './heatmap'
import { add } from './add'

export class Timeline {
  constructor(entries) {
    this.effects = List(entries)
  }

  heatmap() {
    return new Heatmap(
      this.effects.map(heatmap)
                  .reduce(add)
    );
  }
}